hazelcast-client/src/main/java/com/hazelcast/client/config/YamlClientDomConfigProcessor.java

```diff
@@ -0,0 +1,289 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.config;
+
+import com.hazelcast.config.ClassFilter;
+import com.hazelcast.config.DiscoveryConfig;
+import com.hazelcast.config.GlobalSerializerConfig;
+import com.hazelcast.config.ListenerConfig;
+import com.hazelcast.config.SerializationConfig;
+import com.hazelcast.config.SerializerConfig;
+import com.hazelcast.config.yaml.ElementAdapter;
+import com.hazelcast.internal.yaml.YamlMapping;
+import com.hazelcast.internal.yaml.YamlNode;
+import com.hazelcast.internal.yaml.YamlScalar;
+import org.w3c.dom.NamedNodeMap;
+import org.w3c.dom.Node;
+
+import java.nio.ByteOrder;
+import java.util.Map;
+import java.util.Properties;
+
+import static com.hazelcast.config.DomConfigHelper.childElements;
+import static com.hazelcast.config.DomConfigHelper.cleanNodeName;
+import static com.hazelcast.config.DomConfigHelper.getBooleanValue;
+import static com.hazelcast.config.DomConfigHelper.getIntegerValue;
+import static com.hazelcast.internal.yaml.YamlUtil.asMapping;
+import static com.hazelcast.internal.yaml.YamlUtil.asScalar;
+
+class YamlClientDomConfigProcessor extends ClientDomConfigProcessor {
+    YamlClientDomConfigProcessor(boolean domLevel3, ClientConfig clientConfig) {
+        super(domLevel3, clientConfig, new QueryCacheYamlConfigBuilderHelper());
+    }
+
+    @Override
+    protected void handleAttributeNode(Node n) {
+        String attributeName = n.getNodeName();
+        String value = n.getNodeValue();
+        clientConfig.setAttribute(attributeName, value);
+    }
+
+    @Override
+    protected void handleClusterMembers(Node node, ClientNetworkConfig clientNetworkConfig) {
+        for (Node child : childElements(node)) {
+            clientNetworkConfig.addAddress(getTextContent(child));
+        }
+    }
+
+    @Override
+    protected void handleOutboundPorts(Node child, ClientNetworkConfig clientNetworkConfig) {
+        for (Node n : childElements(child)) {
+            String value = getTextContent(n);
+            clientNetworkConfig.addOutboundPortDefinition(value);
+        }
+    }
+
+    @Override
+    protected SerializationConfig parseSerialization(final Node node) {
```

Comment:
Why do you need to override `AbstractDomConfigProcessor#parseSerialization()`? This implementation looks very similar. The same question applies to the other, following methods.

Reply:
There are slight differences in the way that the serialization configuration can be described in the XML and the YAML cases. This is the XML configuration:
```
    <serialization>
        [...]
        <data-serializable-factories>
            <data-serializable-factory factory-id="1">com.hazelcast.examples.DataSerializableFactory
            </data-serializable-factory>
        </data-serializable-factories>
        <portable-factories>
            <portable-factory factory-id="2">com.hazelcast.examples.PortableFactory</portable-factory>
        </portable-factories>
        <serializers>
            <global-serializer override-java-serialization="true">com.hazelcast.examples.GlobalSerializerFactory
            </global-serializer>
            <serializer type-class="com.hazelcast.examples.DummyType"
                        class-name="com.hazelcast.examples.SerializerFactory"/>
        </serializers>
        [...]
```
, which looks the following in YAML:
```
  serialization:
    [...]
    data-serializable-factories:
      - factory-id: 1
        class-name: com.hazelcast.examples.DataSerializableFactory
    portable-factories:
      - factory-id: 2
        class-name: com.hazelcast.examples.PortableFactory
    global-serializer:
      override-java-serialization: true
      class-name: com.hazelcast.examples.GlobalSerializerFactory
    serializers:
      - type-class: com.hazelcast.examples.DummyType
        class-name: com.hazelcast.examples.SerializerFactory
    [...]
```
Note that in the YAML configuration we need an extra attribute for the `class-name` in the factories configurations and that the `global-serializer` is pulled up by one level to make it uniquely configurable. 

For this given method, we can add a callback method to `AbstractDomConfigProcessor#parseSerialization` like this:
```
    protected SerializationConfig parseSerialization(final Node node) {
            [...]
            } else if ("java-serialization-filter".equals(name)) {
                fillJavaSerializationFilter(child, serializationConfig);
            } else {
                handleAdditionalSerializationAttribute(name, child, serializationConfig);
            }
            [...]
    }

    protected void handleAdditionalSerializationAttribute(String name, Node child, SerializationConfig serializationConfig) {
        // NOP, meant to be overridden
    }
```
and we can override it in the `YamlClientDomConfigProcessor`:
```
    @Override
    protected void handleAdditionalSerializationAttribute(String name, Node child, SerializationConfig serializationConfig) {
        if ("global-serializer".equals(name)) {
            fillGlobalSerializer(child, serializationConfig);
        }
    }
```
I'm not sure if it's better to read. This approach would come with many small methods, where parts of handling the same node spanned over multiple classes. I felt this is against the cohesion, so made the trade-off to keep some duplication instead of the above approach.

Reply:
Ok, let's leave it this way. Yeah...it always comes to the balance between avoiding the duplications and keeping the readability...